### ISSUE TITLE
Behavior of exit path on 1.10.6 changed for bad

### ISSUE THREAD

**Issue by leleobhz** [zerotier/ZeroTierOne#1940]
Hello!

I have a VPN concentrator (I use Zerotier to access and route traffic to bunch of other VPNs like OpenVPN) and after change from 1.10.3 to 1.10.6 i noticed my connections stopped working on local network and information about exit IP on my.zerotier.com now points to Ips of any VPNs available, but not mine.

I already tried to blacklist VPN prefixes, configuring custom balancer with only my exit interface and even interface blacklist for vtuns, buty notting worked well. Apparently ZT ignores default route for some reason and try to use other interface even without route to internet over there. 

I suspect something on commit https://github.com/zerotier/ZeroTierOne/commit/ce9330e9bb06afc7047fd016d4f9882f4288b9c1 may be affecting this routing election.

What is possible to do in this case?

**Comment by laduke** [comment]
Thanks for writing. A basic full tunnel mod is still working for me. 

Can you check `zerotier-cli info -j` for your listeningAddresses and surfaceAddresses? is there any difference? Does your routing table look different than from what 1.10.3 was creating? What operating system?

**Comment by leleobhz** [comment]
Hello @laduke 

The only difference my `listeningAddresses` and `surfaceAddresses` have is my public IP address.

At this machine, I do not receive routes neither addresses - instead I set it manually - so there is no routing changes.

A thing I noticed is while I ping one machine to another over ZT, outside ZT interface I see traffic for both machines and in both directions, compatible com a echo request and a echo reply, but if I capture on interface, at machine noone can ping, I see both icmp getting in and out, but on other machines I dont see the awnser. 

Also happens with NDP and ARP - affected machine see both in and out traffic, but clients of this server, does not.

This server is a VyOS running zerotier with podman. Following is the podman inspect zerotier:

```[
     {
          "Id": "a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df",
          "Created": "2023-03-27T11:29:17.245650061-03:00",
          "Path": "/entrypoint.sh",
          "Args": [
               "/entrypoint.sh"
          ],
          "State": {
               "OciVersion": "1.0.2-dev",
               "Status": "running",
               "Running": true,
               "Paused": false,
               "Restarting": false,
               "OOMKilled": false,
               "Dead": false,
               "Pid": 146697,
               "ConmonPid": 146694,
               "ExitCode": 0,
               "Error": "",
               "StartedAt": "2023-03-27T11:29:17.339474241-03:00",
               "FinishedAt": "0001-01-01T00:00:00Z",
               "Health": {
                    "Status": "",
                    "FailingStreak": 0,
                    "Log": null
               },
               "CgroupPath": "/machine.slice/libpod-a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df.scope",
               "CheckpointedAt": "0001-01-01T00:00:00Z",
               "RestoredAt": "0001-01-01T00:00:00Z"
          },
          "Image": "0ba99c5d9009da2041bb6d5fe82ad7869a56c99c71af7fb2434217b9b4763df9",
          "ImageDigest": "sha256:591e4772efa17e1a581039cbadaffde03d02e1f58a0d8add1b64d56a2cf99dab",
          "ImageName": "docker.io/zerotier/zerotier:latest",
          "Rootfs": "",
          "Pod": "",
          "ResolvConfPath": "/overlay-containers/a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df/userdata/resolv.conf",
          "HostnamePath": "/overlay-containers/a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df/userdata/hostname",
          "HostsPath": "/overlay-containers/a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df/userdata/hosts",
          "StaticDir": "/usr/lib/live/mount/persistence/container/storage/overlay-containers/a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df/userdata",
          "OCIConfigPath": "/usr/lib/live/mount/persistence/container/storage/overlay-containers/a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df/userdata/config.json",
          "OCIRuntime": "crun",
          "ConmonPidFile": "/run/vyos-container-zerotier.service.pid",
          "PidFile": "/overlay-containers/a2538751863af67cd72a26cf446f9ff3bef980e539fb4e036e580877c3fe01df/userdata/pidfile",
          "Name": "zerotier",
          "RestartCount": 0,
          "Driver": "overlay",
          "MountLabel": "",
          "ProcessLabel": "",
          "AppArmorProfile": "",
          "EffectiveCaps": [
               "CAP_CHOWN",
               "CAP_DAC_OVERRIDE",
               "CAP_FOWNER",
               "CAP_FSETID",
               "CAP_KILL",
               "CAP_NET_ADMIN",
               "CAP_NET_BIND_SERVICE",
               "CAP_SETFCAP",
               "CAP_SETGID",
               "CAP_SETPCAP",
               "CAP_SETUID",
               "CAP_SYS_CHROOT"
          ],
          "BoundingCaps": [
               "CAP_CHOWN",
               "CAP_DAC_OVERRIDE",
               "CAP_FOWNER",
               "CAP_FSETID",
               "CAP_KILL",
               "CAP_NET_ADMIN",
               "CAP_NET_BIND_SERVICE",
               "CAP_SETFCAP",
               "CAP_SETGID",
               "CAP_SETPCAP",
               "CAP_SETUID",
               "CAP_SYS_CHROOT"
          ],
          "ExecIDs": [],
          "GraphDriver": {
               "Name": "overlay",
               "Data": {
                    "LowerDir": "/usr/lib/live/mount/persistence/container/storage/overlay/cea00629e9db075cee1d1c35dbb98f973bba0362a9ba71f769333e290b486710/diff:/usr/lib/live/mount/persistence/container/storage/overlay/a0e8657b57cff73cd1234b12953e13746b864511b2c6cf02f70ff33b358ea453/diff:/usr/lib/live/mount/persistence/container/storage/overlay/a9b8a900706b029892d7984ce90c0dcefdbc004fd0808adf4e761fde970eb46d/diff:/usr/lib/live/mount/persistence/container/storage/overlay/4e6c1a1827c2b497a5429c295f2bc7528c921d6de1989dffd216f19372bd6d7c/diff:/usr/lib/live/mount/persistence/container/storage/overlay/4a4c146c8da13914247df5931eb71ca9a9eefc30ff5f30a5b918391795f080ee/diff:/usr/lib/live/mount/persistence/container/storage/overlay/c7ad03b4218205852c748e1de5ea8288619e320b4cba58764de78ad92e939d79/diff:/usr/lib/live/mount/persistence/container/storage/overlay/ec09eb83ea031896df916feb3a61cefba9facf449c8a55d88667927538dca2b4/diff",
                    "UpperDir": "/usr/lib/live/mount/persistence/container/storage/overlay/cf731786b4e09b2c0d8a538731ab82668da81f80602f440f40c9560832f0f32c/diff",
                    "WorkDir": "/usr/lib/live/mount/persistence/container/storage/overlay/cf731786b4e09b2c0d8a538731ab82668da81f80602f440f40c9560832f0f32c/work"
               }
          },
          "Mounts": [
               {
                    "Type": "bind",
                    "Source": "/config/user-data/zerotier/devicemap",
                    "Destination": "/var/lib/zerotier-one/devicemap",
                    "Driver": "",
                    "Mode": "",
                    "Options": [
                         "rbind"
                    ],
                    "RW": false,
                    "Propagation": "rprivate"
               },
               {
                    "Type": "bind",
                    "Source": "/config/user-data/zerotier/local.conf",
                    "Destination": "/var/lib/zerotier-one/local.conf",
                    "Driver": "",
                    "Mode": "",
                    "Options": [
                         "rbind"
                    ],
                    "RW": false,
                    "Propagation": "rprivate"
               },
               {
                    "Type": "bind",
                    "Source": "/config/user-data/zerotier/<REMOVED_FOR_GITHUB>.local.conf",
                    "Destination": "/var/lib/zerotier-one/networks.d/<REMOVED_FOR_GITHUB>.local.conf",
                    "Driver": "",
                    "Mode": "",
                    "Options": [
                         "rbind"
                    ],
                    "RW": true,
                    "Propagation": "rprivate"
               }
          ],
          "Dependencies": [],
          "NetworkSettings": {
               "EndpointID": "",
               "Gateway": "",
               "IPAddress": "",
               "IPPrefixLen": 0,
               "IPv6Gateway": "",
               "GlobalIPv6Address": "",
               "GlobalIPv6PrefixLen": 0,
               "MacAddress": "",
               "Bridge": "",
               "SandboxID": "",
               "HairpinMode": false,
               "LinkLocalIPv6Address": "",
               "LinkLocalIPv6PrefixLen": 0,
               "Ports": {},
               "SandboxKey": ""
          },
          "Namespace": "",
          "IsInfra": false,
          "IsService": false,
          "Config": {
               "Hostname": "stargate-sg1",
               "Domainname": "",
               "User": "",
               "AttachStdin": false,
               "AttachStdout": false,
               "AttachStderr": false,
               "Tty": true,
               "OpenStdin": true,
               "StdinOnce": false,
               "Env": [
                    "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                    "TERM=xterm",
                    "container=podman",
                    "ZEROTIER_API_SECRET=<REMOVED_FOR_GITHUB>",
                    "ZEROTIER_IDENTITY_PUBLIC=<REMOVED_FOR_GITHUB>",
                    "ZEROTIER_IDENTITY_SECRET=<REMOVED_FOR_GITHUB>",
                    "ZEROTIER_JOIN_NETWORKS=<REMOVED_FOR_GITHUB>",
                    "HOME=/root",
                    "HOSTNAME=stargate-sg1"
               ],
               "Cmd": null,
               "Image": "docker.io/zerotier/zerotier:latest",
               "Volumes": null,
               "WorkingDir": "/",
               "Entrypoint": "/entrypoint.sh",
               "OnBuild": null,
               "Labels": {
                    "PODMAN_SYSTEMD_UNIT": "vyos-container-zerotier.service"
               },
               "Annotations": {
                    "io.container.manager": "libpod",
                    "io.kubernetes.cri-o.Created": "2023-03-27T11:29:17.245650061-03:00",
                    "io.kubernetes.cri-o.TTY": "true",
                    "io.podman.annotations.autoremove": "FALSE",
                    "io.podman.annotations.cid-file": "/run/vyos-container-zerotier.service.cid",
                    "io.podman.annotations.init": "FALSE",
                    "io.podman.annotations.privileged": "FALSE",
                    "io.podman.annotations.publish-all": "FALSE",
                    "org.opencontainers.image.stopSignal": "15"
               },
               "StopSignal": 15,
               "HealthcheckOnFailureAction": "none",
               "CreateCommand": [
                    "/usr/bin/podman",
                    "run",
                    "--conmon-pidfile",
                    "/run/vyos-container-zerotier.service.pid",
                    "--cidfile",
                    "/run/vyos-container-zerotier.service.cid",
                    "--cgroups=no-conmon",
                    "--detach",
                    "--interactive",
                    "--tty",
                    "--replace",
                    "--cap-add=NET_ADMIN",
                    "--memory",
                    "0m",
                    "--shm-size",
                    "64m",
                    "--memory-swap",
                    "0",
                    "--restart",
                    "always",
                    "--name",
                    "zerotier",
                    "--device=/dev/net/tun:/dev/net/tun",
                    "--volume",
                    "/config/user-data/zerotier/<REMOVED_FOR_GITHUB>.local.conf:/var/lib/zerotier-one/networks.d/<REMOVED_FOR_GITHUB>.local.conf:rw",
                    "--volume",
                    "/config/user-data/zerotier/devicemap:/var/lib/zerotier-one/devicemap:ro",
                    "--volume",
                    "/config/user-data/zerotier/local.conf:/var/lib/zerotier-one/local.conf:ro",
                    "--env",
                    "ZEROTIER_API_SECRET=<REMOVED_FOR_GITHUB>",
                    "--env",
                    "ZEROTIER_IDENTITY_PUBLIC=<REMOVED_FOR_GITHUB>",
                    "--env",
                    "ZEROTIER_IDENTITY_SECRET=<REMOVED_FOR_GITHUB>",
                    "--env",
                    "ZEROTIER_JOIN_NETWORKS=<REMOVED_FOR_GITHUB>",
                    "--net",
                    "host",
                    "docker.io/zerotier/zerotier:latest"
               ],
               "Umask": "0022",
               "Timeout": 0,
               "StopTimeout": 10,
               "Passwd": true,
               "sdNotifyMode": "container"
          },
          "HostConfig": {
               "Binds": [
                    "/config/user-data/zerotier/devicemap:/var/lib/zerotier-one/devicemap:ro,rprivate,rbind",
                    "/config/user-data/zerotier/local.conf:/var/lib/zerotier-one/local.conf:ro,rprivate,rbind",
                    "/config/user-data/zerotier/<REMOVED_FOR_GITHUB>.local.conf:/var/lib/zerotier-one/networks.d/<REMOVED_FOR_GITHUB>.local.conf:rw,rprivate,rbind"
               ],
               "CgroupManager": "systemd",
               "CgroupMode": "private",
               "ContainerIDFile": "/run/vyos-container-zerotier.service.cid",
               "LogConfig": {
                    "Type": "journald",
                    "Config": null,
                    "Path": "",
                    "Tag": "",
                    "Size": "0B"
               },
               "NetworkMode": "host",
               "PortBindings": {},
               "RestartPolicy": {
                    "Name": "always",
                    "MaximumRetryCount": 0
               },
               "AutoRemove": false,
               "VolumeDriver": "",
               "VolumesFrom": null,
               "CapAdd": [
                    "CAP_NET_ADMIN"
               ],
               "CapDrop": [
                    "CAP_AUDIT_WRITE",
                    "CAP_MKNOD",
                    "CAP_NET_RAW"
               ],
               "Dns": [],
               "DnsOptions": [],
               "DnsSearch": [],
               "ExtraHosts": [],
               "GroupAdd": [],
               "IpcMode": "shareable",
               "Cgroup": "",
               "Cgroups": "default",
               "Links": null,
               "OomScoreAdj": 0,
               "PidMode": "private",
               "Privileged": false,
               "PublishAllPorts": false,
               "ReadonlyRootfs": false,
               "SecurityOpt": [],
               "Tmpfs": {},
               "UTSMode": "private",
               "UsernsMode": "",
               "ShmSize": 67108864,
               "Runtime": "oci",
               "ConsoleSize": [
                    0,
                    0
               ],
               "Isolation": "",
               "CpuShares": 0,
               "Memory": 0,
               "NanoCpus": 0,
               "CgroupParent": "",
               "BlkioWeight": 0,
               "BlkioWeightDevice": null,
               "BlkioDeviceReadBps": null,
               "BlkioDeviceWriteBps": null,
               "BlkioDeviceReadIOps": null,
               "BlkioDeviceWriteIOps": null,
               "CpuPeriod": 0,
               "CpuQuota": 0,
               "CpuRealtimePeriod": 0,
               "CpuRealtimeRuntime": 0,
               "CpusetCpus": "",
               "CpusetMems": "",
               "Devices": [
                    {
                         "PathOnHost": "/dev/net/tun",
                         "PathInContainer": "/dev/net/tun",
                         "CgroupPermissions": ""
                    }
               ],
               "DiskQuota": 0,
               "KernelMemory": 0,
               "MemoryReservation": 0,
               "MemorySwap": 0,
               "MemorySwappiness": -1,
               "OomKillDisable": false,
               "PidsLimit": 2048,
               "Ulimits": [
                    {
                         "Name": "RLIMIT_NOFILE",
                         "Soft": 1048576,
                         "Hard": 1048576
                    },
                    {
                         "Name": "RLIMIT_NPROC",
                         "Soft": 4194304,
                         "Hard": 4194304
                    }
               ],
               "CpuCount": 0,
               "CpuPercent": 0,
               "IOMaximumIOps": 0,
               "IOMaximumBandwidth": 0,
               "CgroupConf": null
          }
     }
]
```

This server worked well at the point of my PR https://github.com/zerotier/ZeroTierOne/pull/1871 so this is the reason I suspect of any commit after my PR got merged. Since path learning got changed at https://github.com/zerotier/ZeroTierOne/commit/ce9330e9bb06afc7047fd016d4f9882f4288b9c1 - I tried to check if this is reasonable linked to my issue - but I dont know ZT code in depth to check this.

I can also dump packets at both sides but if you want it, i ask you to open a non-public way to send data - since its my personal work network.

**Comment by leleobhz** [comment]
About both machines `zerotier-cli info -j`

VyOS:

```
leonardo@stargate-sg1:~$ show version 
Version:          VyOS 1.4-rolling-202302060317
Release train:    current
...
leonardo@stargate-sg1:~$ sudo podman exec -it zerotier zerotier-cli info -j
WARN[0000] Switching default driver from overlay2 to the equivalent overlay driver 
{
 "address": "<REMOVED_FOR_GITHUB>",
 "clock": 1679953208809,
 "config": {
  "physical": {
   "10.0.5.0/24": {
    "blacklist": true
   },
   "10.0.7.2/24": {
    "blacklist": true
   },
   "10.0.8.0/24": {
    "blacklist": true
   },
   "10.127.1.0/24": {
    "blacklist": true
   },
   "192.168.190.0/24": {
    "blacklist": true
   },
   "192.168.42.0/24": {
    "blacklist": false,
    "trustedPathId": 1
   },
   "192.168.89.0/24": {
    "blacklist": true
   }
  },
  "settings": {
   "allowTcpFallbackRelay": true,
   "defaultBondingPolicy": "v6-prio",
   "forceTcpRelay": false,
   "interfacePrefixBlacklist": [
    "vtun"
   ],
   "listeningOn": [
    "<PRIVATE_IPV4_ADDR>/9993",
    "<PRIVATE_IPV4_ADDR>/20641",
    "<PRIVATE_IPV4_ADDR>/62179",
    "<PUBLIC_IPV6_ADDR>/9993",
    "<PUBLIC_IPV6_ADDR>/20641",
    "<PUBLIC_IPV6_ADDR>/62179"
   ],
   "policies": {
    "v6-prio": {
     "basePolicy": "active-backup",
     "links": {
      "eth0": {
       "ipvPref": 64,
       "mode": "primary"
      }
     }
    }
   },
   "portMappingEnabled": true,
   "primaryPort": 9993,
   "secondaryPort": 0,
   "softwareUpdate": "disable",
   "softwareUpdateChannel": "release",
   "surfaceAddresses": [
    "<PUBLIC_IPV4_ADDR>/20641",
    "<PUBLIC_IPV4_ADDR>/9993",
    "<PUBLIC_IPV6_ADDR>/20641",
    "<PUBLIC_IPV6_ADDR>/9993",
    "<PUBLIC_IPV6_ADDR>/62179",
    "<PRIVATE_IPV4_ADDR>/62179",
    "<PRIVATE_IPV4_ADDR>/20641",
    "<PRIVATE_IPV4_ADDR>/9993"
   ],
   "tertiaryPort": 0
  }
 },
 "online": true,
 "planetWorldId": <REMOVED_FOR_GITHUB>,
 "planetWorldTimestamp": 1644592324813,
 "publicIdentity": "<REMOVED_FOR_GITHUB>",
 "tcpFallbackActive": false,
 "version": "1.10.6",
 "versionBuild": 0,
 "versionMajor": 1,
 "versionMinor": 10,
 "versionRev": 6
}

```

A note about this blacklist: Another new issue with Zerotier 1.10.6: All this interfaces are OpenVPN tunnels of my customers (I'm a solo IT consultant) and ZT started to FORCE these paths. Until I remove blacklist, my.zerotier.com only shown my customers clients via VPN instead to use my network - even with UPnP and IPv6 properly configured. I also configured a interface blacklist for vtun for same reason.

A client machine (On same private network):

```
 root  ~  screenfetch -nN                                                                                                                                                                                                           SIGINT 
 root@miriam
 OS: Fedora 
 Kernel: x86_64 Linux 6.2.7-200.fc37.x86_64
 Uptime: 9h 24m
 Packages: 2346
 Shell: bash
 Resolution: No X Server
 WM: Not Found
 GTK Theme: Adwaita [GTK3]
 Disk: 78G / 222G (36%)
 CPU: Intel Core i7-7820X @ 16x 4.3GHz [48.0°C]
 GPU: NVIDIA GeForce GT 1030, NVIDIA GeForce RTX 2060
 RAM: 6261MiB / 31767MiB
 root  ~  zerotier-cli info -j
{
 "address": "<REMOVED_FOR_GITHUB>",
 "clock": 1679953537412,
 "config": {
  "settings": {
   "allowTcpFallbackRelay": true,
   "forceTcpRelay": false,
   "listeningOn": [
    "<PRIVATE_IPV4_ADDR/9993",
    "<PRIVATE_IPV4_ADDR/58908",
    "<PRIVATE_IPV4_ADDR/25797",
    "<PUBLIC_IPV6_ADDR>/9993",
    "<PUBLIC_IPV6_ADDR>/58908",
    "<PUBLIC_IPV6_ADDR>/25797"
   ],
   "portMappingEnabled": true,
   "primaryPort": 9993,
   "secondaryPort": 0,
   "softwareUpdate": "disable",
   "softwareUpdateChannel": "release",
   "surfaceAddresses": [
    "<PUBLIC_IPV6_ADDR>/58908",
    "<PUBLIC_IPV4_ADDR>/58908",
    "<PUBLIC_IPV6_ADDR>/25797",
    "<PUBLIC_IPV6_ADDR>/9993",
    "<PRIVATE_IPV4_ADDR/25797",
    "<PRIVATE_IPV4_ADDR/58908",
    "<PRIVATE_IPV4_ADDR/9993",
    "<PUBLIC_IPV4_ADDR>/10236",
    "<PUBLIC_IPV4_ADDR>/25797"
   ],
   "tertiaryPort": 0
  }
 },
 "online": true,
 "planetWorldId": <REMOVED_FOR_GITHUB>,
 "planetWorldTimestamp": 1644592324813,
 "publicIdentity": "<REMOVED_FOR_GITHUB>",
 "tcpFallbackActive": false,
 "version": "1.10.6",
 "versionBuild": 0,
 "versionMajor": 1,
 "versionMinor": 10,
 "versionRev": 6
}
 root  ~  

```

Also please consider <PUBLIC_IPV6_ADDR> and <PRIVATE_IPV4_ADDR> are machines address and <PUBLIC_IPV4_ADDR> are under NAT.

**Comment by leleobhz** [comment]
Fun fact: Use RPM compiled with ubi9:latest on multistage ubi9-minimal reduced a bit the image-size :)

```
leonardo@stargate-sg1# sudo podman image ls
REPOSITORY                           TAG         IMAGE ID      CREATED         SIZE
localhost/leleobhz/zerotier_no_1914  latest      308087ca618e  20 minutes ago  128 MB
docker.io/zerotier/zerotier          latest      0ba99c5d9009  3 days ago      193 MB
```

**Comment by erikh** [comment]
cheerful golf clap​

impressed.

------- Original Message -------

**Comment by leleobhz** [comment]
> What is now blowing my mind is why `"interfacePrefixBlacklist": [ "vtun" ]` alone did not worked, but only after set a new load balance policy considering only eth0 **and** the `"interfacePrefixBlacklist": [ "vtun" ],`. Also blows my mind why blacklisting VPN endpoint networks also stops flow on non-blacklisted interface.

Hello @laduke and @erikh 

The question about `interfacePrefixBlacklist` may deserve a new issue or may be handled here? At least `interfacePrefixBlacklist` documentation may be clarified if this is the intended behavior or some adjust on source may be needed?

**Comment by laduke** [comment]
I am guessing some multipath feature is making interfacePrefixBlacklist not checked. Or you nee

Adding "interfacePrefixBlacklist" to a local.conf on 1.10.6 linux is still working for me.
